File: src/Database/Python/Tests/collection_test.py
# -*- coding: utf-8 -*-
__copyright__ = """This code is licensed under the 3-clause BSD license.
Copyright ETH Zurich, Department of Chemistry and Applied Biosciences, Reiher Group.
See LICENSE.txt for details.
"""

import scine_database as db
import unittest
import os
from json import dumps


class CollectionTest(unittest.TestCase):
    def setUp(self):
        self.manager = db.Manager()
        self.manager.credentials.hostname = os.environ.get(
            'TEST_MONGO_DB_IP', '127.0.0.1')
        self.manager.credentials.port = int(os.environ.get('TEST_MONGO_DB_PORT', 27017))
        self.manager.credentials.database_name = "unittest_db_CollectionTest"
        self.manager.connect()
        self.manager.init()

    def tearDown(self):
        self.manager.wipe()

    def test_query_id(self):
        coll = self.manager.get_collection("compounds")
        cid = db.ID()
        comp = db.Compound.make([cid], coll)
        # Check
        assert coll.has(comp.get_id())

    def test_return_by_id(self):
        coll = self.manager.get_collection("compounds")
        cid = db.ID()
        comp = db.Compound.make([cid], coll)
        # Check
        result = coll.get_compound(comp.get_id())
        assert comp.get_id() == result.get_id()

    def test_return_by_id_fails(self):
        coll = self.manager.get_collection("compounds")
        cid = db.ID()
        self.assertRaises(RuntimeError, lambda: coll.get_compound(cid))

    def test_query_by_dict(self):
        # Make sure the DB is clean in order to allow for accurate counts.
        self.manager.wipe()
        self.manager.init()
        # Setup
        coll = self.manager.get_collection("compounds")
        id1 = db.ID()
        id2 = db.ID()
        id3 = db.ID()
        comp1 = db.Compound.make([id1], coll)
        comp2 = db.Compound.make([id1], coll)
        _ = db.Compound.make([id2], coll)
        _ = db.Compound.make([id3], coll)

        query = {"structures": {"$eq": {"$oid": id1.string()}}}
        vec1 = coll.query_compounds(dumps(query))
        vec2 = coll.query_structures(dumps(query))

        # Check
        assert len(vec1) == 2
        assert vec1[0].get_id() == comp1.get_id()
        assert vec1[1].get_id() == comp2.get_id()
        assert len(vec2) == 0

    def test_get_one(self):
        # Make sure the DB is clean in order to allow for accurate counts.
        self.manager.wipe()
        self.manager.init()
        # Setup
        coll = self.manager.get_collection("compounds")
        id1 = db.ID()
        _ = db.ID()
        id3 = db.ID()
        comp1 = db.Compound.make([id1], coll)
        _ = db.Compound.make([id1], coll)
        _ = db.Compound.make([id3], coll)
        # Create a compound that could be found but should not to catch false
        #   positives
        query1 = {"structures": {"$eq": {"$oid": id1.string()}}}
        query2 = {"structures": {"$eq": {"$oid": id3.string()}}}
        r1 = coll.get_one_compound(dumps(query1))
        r2 = coll.get_one_compound(dumps(query2))

        # Check
        assert r1.has_id()
        assert r1.id() == comp1.id()
        assert r2.has_id()
        self.assertRaises(RuntimeError, lambda: coll.get_one_structure(dumps(query1)))

    def get_one_with_sort(self):
        self.manager.wipe()
        self.manager.init()

        # Setup
        coll = self.manager.get_collection("compounds")
        id1 = db.ID()
        id2 = db.ID()

        comp1 = db.Compound.make([id1], coll)
        comp2 = db.Compound.make([id2, id1], coll)

        query = {"structures": {"$all": [{"$oid": id1.string()}]}}
        sort1 = {"_id": 1}
        sort2 = {"_id": -1}
        r1 = coll.get_one_compound(query)
        r2 = coll.get_one_compound(query, sort1)
        r3 = coll.get_one_compound(query, sort2)

        assert r1.has_id()
        assert r2.has_id()
        assert r3.has_id()

        assert r1.id() == comp1.id()
        assert r2.id() == comp1.id()
        assert r3.id() == comp2.id()

        assert r1.get_structures()[0] == id1
        assert r2.get_structures()[0] == id1
        assert r3.get_structures()[0] == id2
        self.assertRaises(RuntimeError, lambda: coll.get_one_structure(dumps(query)))

    def get_one_and_modify(self):
        self.manager.wipe()
        self.manager.init()

        coll = self.manager.get_collection("compounds")
        id1 = db.ID()
        id2 = db.ID()
        id3 = db.ID()
        id4 = db.ID()
        comp1 = db.Compound.make([id1], coll)
        _ = db.Compound.make([id1], coll)
        _ = db.Compound.make([id3], coll)

        query1 = {"structures": {"$eq": {"$oid": id1.string()}}}
        query2 = {"structures": {"$eq": {"$oid": id2.string()}}}
        update = {"$set": {"reactions": [{"$oid": id4.string()}]}}
        r1 = coll.get_and_update_one_compound(query1, update)
        r2 = coll.get_and_update_one_compound(query2, update)

        assert r1.has_id()
        assert r1.id() == comp1.id()
        assert r1.get_reactions()[0] == id4
        assert not r2.has_id()
        self.assertRaises(RuntimeError, lambda: coll.get_and_update_one_structure(dumps(query1, update)))

    def get_one_and_modify_with_sort(self):
        self.manager.wipe()
        self.manager.init()

        coll = self.manager.get_collection("compounds")
        id1 = db.ID()
        id2 = db.ID()
        id3 = db.ID()
        _ = db.Compound.make([id1], coll)
        comp2 = db.Compound.make([id2, id1], coll)

        query = {"structures": {"$all": [{"$oid": id1.string()}]}}
        update = {"$set": {"reactions": [{"$oid": id3.string()}]}}
        sort = {"_id": -1}
        r1 = coll.get_and_update_one_compound(query, update, sort)
        assert r1.has_id()
        assert r1.id() == comp2.id()
        assert r1.get_reactions()[0] == id3
        self.assertRaises(RuntimeError, lambda: coll.get_one_structure(dumps(query, update, sort)))

    def test_random_select(self):
        # Make sure the DB is clean in order to allow for accurate counts.
        self.manager.wipe()
        self.manager.init()
        # Setup
        coll = self.manager.get_collection("compounds")
        id1 = db.ID()
        id2 = db.ID()
        id3 = db.ID()
        _ = db.Compound.make([id1], coll)
        _ = db.Compound.make([id2], coll)
        _ = db.Compound.make([id3], coll)

        # Check correct number
        nSamples = 2
        vec1 = coll.random_select_compounds(nSamples)
        assert len(vec1) == nSamples

        # Check if higher number is possible
        nSamples = 10
        vec2 = coll.random_select_compounds(nSamples)
        assert len(vec2) != nSamples
        assert len(vec2) == coll.count("{}")

    def test_count_by_dict(self):
        # Make sure the DB is clean in order to allow for accurate counts.
        self.manager.wipe()
        self.manager.init()
        # Setup
        coll = self.manager.get_collection("compounds")
        id1 = db.ID()
        id2 = db.ID()
        _ = db.Compound.make([id1], coll)
        _ = db.Compound.make([id1], coll)
        _ = db.Compound.make([id2], coll)

        query = {"structures": {"$eq": {"$oid": id1.string()}}}
        result = coll.count(dumps(query))

        # Check
        assert result == 2

    def test_loop(self):
        # Make sure the DB is clean in order to allow for accurate counts.
        self.manager.wipe()
        self.manager.init()
        # Setup
        coll = self.manager.get_collection("compounds")
        id1 = db.ID()
        id2 = db.ID()
        _ = db.Compound.make([id1], coll)
        _ = db.Compound.make([id1], coll)
        _ = db.Compound.make([id2], coll)

        query = {"structures": {"$eq": {"$oid": id1.string()}}}

        count = 0
        for comp in coll.iterate_compounds(dumps(query)):
            count += 1
            comp.link(coll)
            assert comp.has_structure(id1)

        # Check
        assert count == 2

File: src/Database/Tests/CollectionTest.cpp
/**
 * @file
 * @copyright This code is licensed under the 3-clause BSD license.\n
 *            Copyright ETH Zurich, Department of Chemistry and Applied Biosciences, Reiher Group.\n
 *            See LICENSE.txt for details.
 */
#include <Database/Collection.h>
#include <Database/Exceptions.h>
#include <Database/Manager.h>
#include <Database/Objects/Compound.h>
#include <Database/Objects/Structure.h>
#include <gmock/gmock.h>

using namespace testing;
namespace Scine {
namespace Database {
namespace Tests {
/**
 * @class CollectionTest CollectionTest.cpp
 * @brief Comprises tests for the class Scine::Database::Collection.
 * @test
 */
struct CollectionTest : public Test {
  static Credentials credentials;
  static Manager db;

  static void SetUpTestCase() {
    db.setCredentials(credentials);
    db.connect();
    db.init();
  }
  static void TearDownTestCase() {
    db.wipe();
  }
};

Credentials CollectionTest::credentials(TEST_MONGO_DB_IP, std::atoi(TEST_MONGO_DB_PORT), "unittest_db_CollectionTest");
Manager CollectionTest::db;

TEST_F(CollectionTest, QueryID) {
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1;
  Compound comp = Compound::create({id1}, coll);

  // Check
  ASSERT_TRUE(coll->has(comp.id()));
}

TEST_F(CollectionTest, ReturnObjectByID) {
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1;
  Compound comp = Compound::create({id1}, coll);

  // Get
  auto result = coll->get<Compound>(comp.id());
  ASSERT_EQ(result.id(), comp.id());
}

TEST_F(CollectionTest, ReturnObjectByIDFails) {
  auto coll = db.getCollection("compounds");
  ID id1;
  ASSERT_THROW(coll->get<Compound>(id1), Exceptions::IDNotFoundException);
}

TEST_F(CollectionTest, QueryByJSON) {
  // Make sure the DB is clean in order to allow for accurate counts.
  db.wipe();
  db.init();
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1, id2, id3;
  Compound comp1 = Compound::create({id1}, coll);
  Compound comp2 = Compound::create({id1}, coll);
  Compound comp3 = Compound::create({id3}, coll);

  std::string query = R"({ "structures" : { "$eq" : { "$oid" : ")" + id1.string() + "\" } } }";
  auto vec1 = coll->query<Compound>(query);
  auto vec2 = coll->query<Structure>(query);

  // Check
  ASSERT_EQ(vec1.size(), 2);
  ASSERT_EQ(vec1[0].id(), comp1.id());
  ASSERT_EQ(vec1[1].id(), comp2.id());
  ASSERT_EQ(vec2.size(), 0);
}

TEST_F(CollectionTest, GetOne) {
  // Make sure the DB is clean in order to allow for accurate counts.
  db.wipe();
  db.init();
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1, id2, id3;
  Compound comp1 = Compound::create({id1}, coll);
  Compound comp2 = Compound::create({id1}, coll);
  Compound comp3 = Compound::create({id3}, coll);
  // Create a compound that could be found but should not to catch false
  // positives

  const std::string query1 = R"({ "structures" : { "$eq" : { "$oid" : ")" + id1.string() + "\" } } }";
  const std::string query2 = R"({ "structures" : { "$eq" : { "$oid" : ")" + id2.string() + "\" } } }";
  auto r1 = coll->getOne<Compound>(query1);
  auto r2 = coll->getOne<Compound>(query2);

  // Check
  ASSERT_TRUE(r1.hasId());
  ASSERT_TRUE(r1.id() == comp1.id());
  ASSERT_FALSE(r2.hasId());
  ASSERT_THROW(coll->getOne<Structure>(query1), Exceptions::ObjectTypeMismatchException);
}

TEST_F(CollectionTest, GetOneWithSort) {
  // Make sure the DB is clean in order to allow for accurate counts.
  db.wipe();
  db.init();
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1, id2;
  Compound comp1 = Compound::create({id1}, coll);
  Compound comp2 = Compound::create({id2, id1}, coll);

  const std::string query = R"({ "structures" : { "$all" : [{ "$oid" : ")" + id1.string() + "\" }] } }";
  const std::string sort1 = R"({"_id": 1})";
  const std::string sort2 = R"({"_id": -1})";
  auto r1 = coll->getOne<Compound>(query);
  auto r2 = coll->getOne<Compound>(query, sort1);
  auto r3 = coll->getOne<Compound>(query, sort2);

  ASSERT_TRUE(r1.hasId());
  ASSERT_TRUE(r2.hasId());
  ASSERT_TRUE(r3.hasId());

  ASSERT_TRUE(r1.id() == comp1.id());
  ASSERT_TRUE(r2.id() == comp1.id());
  ASSERT_TRUE(r3.id() == comp2.id());

  ASSERT_TRUE(r1.getStructures()[0] == id1);
  ASSERT_TRUE(r2.getStructures()[0] == id1);
  ASSERT_TRUE(r3.getStructures()[0] == id2);

  ASSERT_THROW(coll->getOne<Structure>(query, sort1), Exceptions::ObjectTypeMismatchException);
}

TEST_F(CollectionTest, GetOneAndModify) {
  // Make sure the DB is clean in order to allow for accurate counts.
  db.wipe();
  db.init();
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1, id2, id3, id4;
  Compound comp1 = Compound::create({id1}, coll);
  Compound comp2 = Compound::create({id1}, coll);
  Compound comp3 = Compound::create({id3}, coll);
  // Create a compound that could be found but should not to catch false
  // positives

  const std::string query1 = R"({ "structures" : { "$eq" : { "$oid" : ")" + id1.string() + "\" } } }";
  const std::string query2 = R"({ "structures" : { "$eq" : { "$oid" : ")" + id2.string() + "\" } } }";
  const std::string update = R"({ "$set" : { "reactions" : [{ "$oid" : ")" + id4.string() + "\" }] } }";
  auto r1 = coll->getAndUpdateOne<Compound>(query1, update);
  auto r2 = coll->getAndUpdateOne<Compound>(query2, update);

  // Check
  ASSERT_TRUE(r1.hasId());
  ASSERT_TRUE(r1.id() == comp1.id());
  ASSERT_TRUE(r1.getReactions()[0] == id4);
  ASSERT_FALSE(r2.hasId());
  ASSERT_THROW(coll->getAndUpdateOne<Structure>(query1, update), Exceptions::ObjectTypeMismatchException);
}

TEST_F(CollectionTest, GetOneAndModifyWithSort) {
  // Make sure the DB is clean in order to allow for accurate counts.
  db.wipe();
  db.init();
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1, id2, id3;
  Compound comp1 = Compound::create({id1}, coll);
  Compound comp2 = Compound::create({id2, id1}, coll);
  // Create a compound that could be found but should not to catch false
  // positives

  const std::string query = R"({ "structures" : { "$all" : [{ "$oid" : ")" + id1.string() + "\" }] } }";
  const std::string update = R"({ "$set" : { "reactions" : [{ "$oid" : ")" + id3.string() + "\" }] } }";
  const std::string sort = R"({"_id": -1})";
  auto r1 = coll->getAndUpdateOne<Compound>(query, update, sort);

  // Check
  ASSERT_TRUE(r1.hasId());
  ASSERT_TRUE(r1.id() == comp2.id());
  ASSERT_TRUE(r1.getReactions()[0] == id3);

  ASSERT_THROW(coll->getAndUpdateOne<Structure>(query, update, sort), Exceptions::ObjectTypeMismatchException);
}

TEST_F(CollectionTest, RandomSelect) {
  // Make sure the DB is clean in order to allow for accurate counts.
  db.wipe();
  db.init();
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1, id2, id3;
  Compound comp1 = Compound::create({id1}, coll);
  Compound comp2 = Compound::create({id2}, coll);
  Compound comp3 = Compound::create({id3}, coll);

  // Check correct number
  std::int32_t nSamples = 2;
  std::vector<Compound> vec1 = coll->randomSelect<Compound>(nSamples);
  ASSERT_EQ(vec1.size(), nSamples);

  // Check if higher number is possible
  nSamples = 10;
  std::vector<Compound> vec2 = coll->randomSelect<Compound>(nSamples);
  ASSERT_NE(vec2.size(), nSamples);
  ASSERT_EQ(vec2.size(), coll->count("{}"));
}

TEST_F(CollectionTest, CountByJSON) {
  // Make sure the DB is clean in order to allow for accurate counts.
  db.wipe();
  db.init();
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1, id2;
  Compound comp1 = Compound::create({id1}, coll);
  Compound comp2 = Compound::create({id1}, coll);
  Compound comp3 = Compound::create({id2}, coll);

  std::string query = R"({ "structures" : { "$eq" : { "$oid" : ")" + id1.string() + "\" } } }";
  auto result = coll->count(query);

  // Check
  ASSERT_EQ(result, 2);
}

TEST_F(CollectionTest, TestLoop) {
  // Make sure the DB is clean in order to allow for accurate counts.
  db.wipe();
  db.init();
  // Setup
  auto coll = db.getCollection("compounds");
  ID id1, id2;
  Compound comp1 = Compound::create({id1}, coll);
  Compound comp2 = Compound::create({id1}, coll);
  Compound comp3 = Compound::create({id2}, coll);

  std::string query = R"({ "structures" : { "$eq" : { "$oid" : ")" + id1.string() + "\" } } }";

  int count = 0;
  for (auto iter = coll->iteratorQuery<Compound>(query); !iter.done(); iter++) {
    count++;
    auto comp = *iter;
    comp.link(coll);
    ASSERT_TRUE(comp.hasStructure(id1));
  }

  // Check
  ASSERT_EQ(count, 2);
}

} /* namespace Tests */
} /* namespace Database */
} /* namespace Scine */
